GDScript(2.0) Documentation generation system

ThakeeNathees opened the pull request:
Suppression of my initial pr(https://github.com/godotengine/godot/pull/39359). Refactored with GDScript 2.0

![new-doc](https://user-images.githubusercontent.com/41085900/89616951-e799ab80-d8a6-11ea-991a-f4012d9b37da.jpg)

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/993.*

AndreaCatania commented:
Good job. If it's not too difficult, would be nice highlight the comment, like happens with syntax `""" Comment """`.

ThakeeNathees commented:
@AndreaCatania IMO, making the comments/documentations less visible from the source is easy to read the code, 

here is a side by side comparison 
![Screenshot (38)](https://user-images.githubusercontent.com/41085900/89705506-99a2a780-d97b-11ea-9dc3-43d9e50e9ced.jpg)

AndreaCatania commented:
Usually the documenting code is marked with a different colour. Probably that yellow is wrong and too strong for its purpose.

dalexeev commented:
Perhaps a dark purple or dark green color will work. They are not used.

vnen commented on editor/editor_file_system.cpp:
```diff
@@ -794,6 +794,17 @@ void EditorFileSystem::_scan_new_dir(EditorFileSystemDirectory *p_dir, DirAccess
 		}
 
+		// TODO-DOC: is it okey to call this here? some how the script
+		//  needs to be compiled to generate and cache the documentations.
```
I'm notsure about this. Will have to test in a big project to see if there are issues or slowdowns. I think this runs in a thread, so it might not be so bad.

vnen reviewed (changes requested):
Overall looks great, just a few things that needs to change.

neikeq commented on modules/mono/csharp_script.h:
```diff
@@ -188,6 +188,13 @@ class CSharpScript : public Script {
+#ifdef TOOLS_ENABLED
+	virtual const Vector<DocData::ClassDoc> &get_documentation() const override {
+		static Vector<DocData::ClassDoc> docs;
+		return docs;
```
We can definitely support this in C# in the future. Could you add a `// TODO` here?

ThakeeNathees replied:
@neikeq Okey!

ThakeeNathees commented on modules/gdscript/gdscript_tokenizer.h:
```diff
@@ -181,6 +182,21 @@ class GDScriptTokenizer {
+			new_line = p_new_line;
+		}
+	};
+	virtual const Map<int, CommentData> &get_comments() const {
```
before GDScript 2.0 `GDScriptTokenizerBuffer` `GDScriptTokenizerText` were supposed to have these method, now I guess I can remove the virtual.

vnen replied:
@ThakeeNathees I don't think we'll split the tokenizer again, and if we do we'll have to change everything anyway. So for now you can remove the `virtual`.

ThakeeNathees replied:
@vnen done.

ThakeeNathees commented:
@vnen completed!

vnen reviewed (approved):
LGTM. Thanks for the work!

akien-mga commented on core/script_language.h:
```diff
@@ -35,6 +35,7 @@
 #include "core/map.h"
 #include "core/pair.h"
 #include "core/resource.h"
+#include "editor/doc_data.h"
```
`core/` should not have dependencies on `editor/` code, so we need to find a different solution here. (At the very list in its current state this should be within `#ifdef TOOLS_ENABLED`, but I think we should find a solution that doesn't break encapsulation instead.)

CC @reduz

akien-mga replied:
A solution might be to move `editor/doc_data.h` to `core/doc_data.h` (to be discussed with @reduz).

akien-mga replied:
13:30 <Akien> reduz: Can you check my review comments at the end of #41095 ? It's pretty solid work from thakeenathees[m] and already approved by vnen, but I have some concerns on the use of `editor/doc_data.h` in `core/script_language.h`
13:30 <Akien> And overall there's a lot of `#ifdef TOOLS_ENABLED` code added in GDScript too which can lead to bugs in tools vs non-tools builds, but I don't really see a better way around it.
13:31 <Akien> I guess doc_data.h should just be moved to core?
13:32 <reduz> I am not sure
13:33 <reduz> it needs scene stuff too
13:33 <reduz> so probably no
13:33 <reduz> maybe we can create an intermediate class with just the structures
13:33 <reduz> and put it in core
13:35 <reduz> another alternative can be to just replicate some of the structures in ScriptLanguage
13:35 <reduz> with an ifdef tools
13:35 <reduz> but it looks like they are the same thing
13:35 <Akien> Seems it only needs scene for `scene/resources/theme.h`, but yeah that's true.
13:35 <reduz> I think best would be to just split it in two
13:35 <reduz> not sure what vnen thinks
13:36 <Akien> Yeah duplicating the struct might be the simplest, only ClassDoc is needed it seems.
13:36 <Akien> (Well ClassDoc depends on all other structs :))
13:36 <reduz> ClassDoc includes all the other structs
13:36 <reduz> yeah, I think the best is to split it
13:38 <Akien> So `core/doc_data.h` with the structs, and `editor/doc_manager.h` with the methods or something like that?
13:39 <Akien> Or DocTools
13:39 <reduz> yeah
```

reduz replied:
This class should be split in two, so you can access doc without having to rely on editor stuff.

ThakeeNathees replied:
since doc_data.h, isn't depend on the editor, I've moved it to the core `core/doc_data.h` and changed all `#include "editor/doc_data.h"` into `#include "core/doc_data.h"` and it's working just fine and I didn't need to split it to an interface and implementation between core and editor, is it okey?
@akien-mga  @reduz

akien-mga replied:
The problem is that `doc_data.cpp` depends on `scene`, and `core` code should not depend on `scene`.

But you can have the data structs in `core` as `DocData`, and the actual utility methods for handling doc data in `editor` as e.g. `DocManager` or `DocTool` (which includes `core/doc_data.h`).

So it would be something like:
```
core/doc_data.h
editor/doc_tools.cpp
editor/doc_tools.h
```

ThakeeNathees replied:
done! I've created a new commit for the doc_data refactor.

akien-mga commented on core/object.h:
```diff
@@ -145,6 +145,9 @@ struct PropertyInfo {
 	PropertyHint hint = PROPERTY_HINT_NONE;
 	String hint_string;
 	uint32_t usage = PROPERTY_USAGE_DEFAULT;
+#ifdef TOOLS_ENABLED
+	String doc_string;
+#endif
```
@reduz Please review this `object.h` addition.

reduz replied:
Please do not add more stuff to these structures, use a different approach to obtain documentation from scripts.

ThakeeNathees replied:
done. descriptions no more depend on it's propertyinfo/methodinfo. (they directly changed to `DocData::*Doc` at compile)

akien-mga commented on editor/doc_data.h:
```diff
@@ -111,8 +129,18 @@ class DocData {
+
 	void merge_from(const DocData &p_data);
 	void remove_from(const DocData &p_data);
+	void add_doc(const ClassDoc &p_class_doc);
+	void remove_doc(const String &p_class_name);
+	bool has_doc(const String &p_class_name);
```
Nitpick, but the argument names differ a bit here.

ThakeeNathees replied:
I match the argument name with it's type add with the `ClassDoc` and remove with it's name

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -506,7 +506,19 @@ void EditorHelp::_update_doc() {
+	if (cd.is_script_doc) {
+		has_properties = false;
+		for (int i = 0; i < cd.properties.size(); i++) {
+			if (cd.properties[i].name.begins_with("_") && cd.properties[i].description == String()) {
```
Not too important, but I'd suggest using `.empty()` instead of ` == String()` to check for empty strings. (It's not done consistently in the codebase currently, we'll have to decide on a convention eventually.)

ThakeeNathees replied:
done!

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -577,6 +593,32 @@ void EditorHelp::_update_doc() {
+
+				if (cd.properties[i].setter != "") {
+					class_desc->push_color(value_color);
+					class_desc->add_text("setter");
```
Should this and `"getter"` use `TTR()`? Not sure how I'd translate them in French though :P

akien-mga replied:
Looks like existing code didn't translate those which are use somewhat like keywords, so I guess maybe it's OK as is.

akien-mga commented on modules/gdnative/nativescript/nativescript.h:
```diff
@@ -152,6 +152,13 @@ class NativeScript : public Script {
 	virtual Error reload(bool p_keep_state = false) override;
 
+#ifdef TOOLS_ENABLED
+	virtual const Vector<DocData::ClassDoc> &get_documentation() const override {
```
You should make the dependency explicit by adding the `#include` (include what you use).
But the problem of the `editor/doc_data.h` dependency in non-editor code needs to be solved first.

ThakeeNathees replied:
done!

akien-mga commented on modules/gdnative/pluginscript/pluginscript_script.h:
```diff
@@ -97,6 +97,13 @@ class PluginScript : public Script {
 	virtual Error load_source_code(const String &p_path);
 
+#ifdef TOOLS_ENABLED
+	virtual const Vector<DocData::ClassDoc> &get_documentation() const override {
```
Same here.

ThakeeNathees replied:
done!

akien-mga commented on modules/gdscript/gdscript.cpp:
```diff
@@ -39,6 +39,9 @@
 #include "core/os/file_access.h"
 #include "core/os/os.h"
 #include "core/project_settings.h"
+#include "editor/doc_tools.h"
+#include "editor/editor_help.h"
+#include "editor/plugins/script_editor_plugin.h"
```
I think there's still an encapsulation issue here, `gdscript.cpp` shouldn't depend on editor code.
Editor-specific GDScript features are in `modules/gdscript/editor/`.

vnen replied:
The editor-specific functions can be implemented in a new file under `editor` (maybe as `gdscript_doc.cpp`). Or you can add to `gdscript_editor.cpp` though that one is a bit cluttered already.

akien-mga replied:
That's the only thing left to rework IMO, then we can merge. If you're not sure how to do it, maybe @vnen can help.

ThakeeNathees replied:
done! I was using script editor in gdscript to update the docs, where I should've used the `Ref<Script>` interface from the script editor.
@akien-mga  @vnen

akien-mga replied:
Awesome! :)

ThakeeNathees replied:
save as... didn't update the helper, and I've just fixed it. if there any edge cases let me know

akien-mga commented on modules/gdscript/gdscript.cpp:
```diff
@@ -640,6 +848,10 @@ Error GDScript::reload(bool p_keep_state) {
 	err = compiler.compile(&parser, this, p_keep_state);
 
+#if TOOLS_ENABLED
+	this->_update_doc();
```
```suggestion
	_update_doc();
```

akien-mga commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2749,6 +2796,11 @@ Error GDScriptCompiler::_parse_class_level(GDScript *p_script, const GDScriptPar
 					parameters_names.write[j] = signal->parameters[j]->identifier->name;
 				}
 				p_script->_signals[name] = parameters_names;
+#if TOOLS_ENABLED
```
```suggestion
#ifdef TOOLS_ENABLED
```

akien-mga commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2674,20 +2707,23 @@ Error GDScriptCompiler::_parse_class_level(GDScript *p_script, const GDScriptPar
 				} else {
 					prop_info.usage = PROPERTY_USAGE_SCRIPT_VARIABLE;
 				}
+#if TOOLS_ENABLED
```
```suggestion
#ifdef TOOLS_ENABLED
```

vnen replied:
This too.

akien-mga commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2575,6 +2590,24 @@ Error GDScriptCompiler::_parse_class_level(GDScript *p_script, const GDScriptPar
 		}
 	}
 
+#if TOOLS_ENABLED
```
```suggestion
#ifdef TOOLS_ENABLED
```

vnen replied:
This is still a problem.

akien-mga replied:
Indeed, total list of the wrong checks:
```
$ rg "if TOOLS"
modules/mono/csharp_script.cpp
2987:#if TOOLS_ENABLED

modules/gdscript/gdscript.cpp
243:#if TOOLS_ENABLED
836:#if TOOLS_ENABLED

modules/gdscript/gdscript_compiler.cpp
2012:#if TOOLS_ENABLED
2145:#if TOOLS_ENABLED
2227:#if TOOLS_ENABLED
```

ThakeeNathees replied:
done!

akien-mga reviewed (changes requested):
Sorry for the late re-review. Seems mostly good but there's one issue left with `gdscript.cpp` now including editor-specific code.

vnen commented:
Hello @ThakeeNathees, very sorry for the delay on this and because of that now it has many conflicts. Are you willing to fix those? We can do a final review then (so far the code looks mostly good to be merged, just the conflicts and minor details are in the way). If you don't want to I can do it. I think it will be easier for you because you're more familiar with your code. I can help if you have doubts about something.

ThakeeNathees commented:
@vnen Thanks, I'll fix those conflicts soon

ThakeeNathees committed:
Documentation generation for GDScript

- ClassDoc added to GDScript and property reflection data were extracted
from parse tree

- GDScript comments are collected from tokenizer for documentation and
applied to the ClassDoc by the GDScript compiler

- private docs were excluded (name with underscore prefix and doesn't
have any doc comments)

- default values (of non exported vars), arguments are extraced from the
parser

- Integrated with GDScript 2.0 and new enums were added.

- merge conflicts fixed

ThakeeNathees committed:
Refactor DocData into core and editor (DocTools) parts

akien-mga commented:
There are some style issues to fix: https://github.com/godotengine/godot/pull/41095/checks?check_run_id=1470115199

Maybe you can also amend the commit message for the second commit to be a bit more explicit, e.g. "Refactor DocData into core and editor (DocTools) parts".

akien-mga commented on modules/gdscript/gdscript.h:
```diff
@@ -33,10 +33,11 @@
 #include "core/io/resource_saver.h"
 #include "core/object/script_language.h"
-#include "editor/doc_data.h"
+#include "core/doc_data.h"
```
You include this twice.

ThakeeNathees replied:
done the merge conflicts and style checks, I fix the  encapsulation issue ASAP

akien-mga reviewed (approved):
Amazing work! I'm eager to try it out :)

All my initial comments from a quick overview are solved. I haven't dug down in the details everywhere but the implementation seems sensible.

I'll let @vnen see if he wants to do a quick last review or merge as is. I'll try to test the code a bit in the meantime.

vnen reviewed (changes requested):
Looks really good. Just a small thing that you missed.

vnen reviewed (approved):
LGTM. Great work!

akien-mga commented:
Thanks hugely for this amazing feature, and for following up on belated reviews after GSoC :)

ithinkandicode commented:
Will this feature ever be available in 3.x?

YuriSizov commented:
@ithinkandicode That's unlikely, as this feature is built on top of the changes in master which are not compatible. Nothing is impossible, but that would require dedicated 3.x GDScript maintainers, and there is currently nobody interested in doing that work.

the-eclectic-dyslexic commented:
I am noticing the tooltips don't update until restarting in 4 beta 11. Is this intended behaviour or am I missing an option to tell them to regenerate?

akien-mga commented:
That's a known issue / expected behavior for any change that impacts an already open inspector for the node you're editing. If you click another node and then back to this one, it should refresh.

the-eclectic-dyslexic commented:
@akien-mga Thanks for the reply, that was helpful. Have a good one!

Aristottle commented:
The tooltips don't seem to generate if the variable is within an export category

EvanHecht commented:
> The tooltips don't seem to generate if the variable is within an export category

Yeah can confirm this happening for me as well.

vnen commented:
Not sure why this is receiving comments lately so I locked to avoid cluttering the discussion. If you have a problem with the feature then open an issue about it since comments in merged pull requests tend to be forgotten.
